[browser][wasm] Wasm websockets support

kjpou1 opened the pull request.

kjpou1 committed:
[browser][wasm] Initial addition of WebSockets support

kjpou1 committed:
Clean up usings

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ClientWebSocket.cs:
```diff
@@ -0,0 +1,563 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Rather than entirely re-implementing ClientWebSocket, can you instead just create a WebSocketHandle.Browser.cs ala the WebSocketHandle.Managed.cs at https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs ?  This "handle" is how we handled (no pun intended) the cross-platformness in the past here, with a WebSocketHandle.WinHttp.cs and a WebSocketHandle.WinRT.cs in previous releases.  We no longer have those because they now just use the managed one, but the abstraction remains in place for need like the one in this PR.

kjpou1 replied:
Hey Stephen.  

We are discussing the same thing and seeing if it actually fits.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Create and use a WebSocketHandle.Browser.cs module

kjpou1 committed:
Add conditional TARGETS_BROWSER so as not to throw Proxy PNSE during ClientWebSocket creation.

kjpou1 committed:
Update WebSockets code

kjpou1 committed:
Removing debug line

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ClientWebSocketOptions.cs:
```diff
@@ -0,0 +1,122 @@
+    public sealed class ClientWebSocketOptions
+    {
+        private bool _isReadOnly; // After ConnectAsync is called the options cannot be modified.
+        internal List<string>? _requestedSubProtocols;
```
Why is it internal?

kjpou1 replied:
not sure to tell you the truth.  I followed the same as here:  https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocketOptions.cs#L27

Can change this one as well as the one in master as well.

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ClientWebSocketOptions.cs:
```diff
@@ -0,0 +1,122 @@
+
+        public void SetBuffer(int receiveBufferSize, int sendBufferSize)
+        {
+            throw new NotImplementedException();
```
Why NotImplementedException ?

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ClientWebSocketOptions.cs:
```diff
@@ -0,0 +1,122 @@
+            _isReadOnly = true;
+        }
+
+        private void ThrowIfReadOnly()
```
Is it worth to have method for a single call?

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ReceivePayload.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace System.Net.WebSockets
+{
+    internal sealed class ReceivePayload
```
Where is this type used?

kjpou1 replied:
Used by BrowserWebSocket.cs to handle the received message queue

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs:
```diff
@@ -31,8 +31,11 @@ public ClientWebSocket()
             WebSocketHandle.CheckPlatformSupport();
 
             _state = (int)InternalState.Created;
+#if TARGETS_BROWSER
```
You could if-def Proxy setting only

kjpou1 replied:
addressed in recent commits

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs:
```diff
@@ -72,6 +73,40 @@ public void Abort()
             _webSocket!.CloseOutputAsync(closeStatus, statusDescription, cancellationToken);
 
+#if TARGETS_BROWSER
+        public async Task ConnectAsyncCore(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
```
As this is public API it does not make much sense to if-def the method declaration

kjpou1 committed:
Remove `ThrowIfReadOnly` method to address review comments

kjpou1 committed:
Address review comments

- change accessor of _requestedSubProtocols to private
- condition more code in WebSocketHandle.Managed.cs to not access _requestedSubProtocols

kjpou1 committed:
Address review comments

kjpou1 committed:
Address review comments

kjpou1 committed:
Change to be PlatformNotSupportedException

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

# Conflicts:
#	src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj
#	src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

kjpou1 commented:
@marek-safar @stephentoub ping for a review when available.

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    subProtocols?.Dispose();
+
+                    // Setup the onError callback
+                    _onError = new Action<JSObject>((errorEvt) =>
```
```suggestion
					  _onError = (errorEvt) => errorEvt.Dispose();
```

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    {
+                        _innerWebSocketCloseStatus = (WebSocketCloseStatus)closeEvt.GetObjectProperty("code");
+                        _innerWebSocketCloseStatusDescription = closeEvt.GetObjectProperty("reason")?.ToString();
+                        var mess = new ReceivePayload(ArraySegment<byte>.Empty, WebSocketMessageType.Close);
```
- Use EmptyArray instead of ArraySegment<byte>.Empty
- change local variable name

kjpou1 replied:
Addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+
+                        if (!tcsConnect.Task.IsCanceled && !tcsConnect.Task.IsCompleted && !tcsConnect.Task.IsFaulted)
+                        {
+                            tcsConnect.SetException(new WebSocketException(WebSocketError.NativeError));
```
When is this going to be hit?

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    // Setup the onOpen callback
+                    _onOpen = new Action<JSObject>((evt) =>
+                    {
+                        using (evt)
```
Why is the using needed ?

kjpou1 replied:
Really it is not needed anymore with the reference counting PR but since it is a native JSObject I prefer to release it as soon as possible.

I can change this if it is preferred.

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                                                using (loadEvent)
+                                                using (var target = (JSObject)loadEvent.GetObjectProperty("target"))
+                                                {
+                                                    if ((int)target.GetObjectProperty("readyState") == 2)
```
What does 2 mean in this context?

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                }
+                catch (Exception wse)
+                {
+                    ConnectExceptionCleanup();
```
Inline the Dispose call

kjpou1 replied:
addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+        {
+            if (State == WebSocketState.Open)
+            {
+                var closeResult = CloseAsyncCore(WebSocketCloseStatus.NormalClosure, "Connection was aborted", CancellationToken.None);
```
Shouldn't this wait for the result?

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    nameof(messageType));
+            }
+
+            if (!endOfMessage && buffer.Array == null)
```
You have same check 3 lines bellow

kjpou1 replied:
Addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                return;
+            }
+            else if (_writeBuffer != null && buffer.Array != null)
```
Redundant else

kjpou1 replied:
Addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            {
+                try
+                {
+                    if (messageType == WebSocketMessageType.Binary)
```
Suggestion, use switch

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                {
+                    var endOfMessage = _bufferedPayload.BufferPayload(buffer, out WebSocketReceiveResult receiveResult);
+
+                    tcsReceive.SetResult(receiveResult);
```
What is the intention of this API, should it set the result buffer before it completes?

@stephentoub

stephentoub replied:
I'm not clear on why the TCS is needed at all.  But assuming it is needed, it seems like the actual read is done by DequeuePayloadAsync, and so this does seem to be setting it at the right time.

kjpou1 replied:
Stephen

The TCS is needed because we do not have threading enabled at this time. The only way to have await work is to use TaskCompletionSource to control when the task should return with the result.

kjpou1 replied:
Addressed as well

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    throw new WebSocketException(WebSocketError.NativeError, exc);
+                }
+
+                return await tcsReceive.Task.ConfigureAwait(continueOnCapturedContext: true);
```
This should be just `return tcsReceive.Task;`

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                _innerWebSocketCloseStatus = closeStatus;
+                _innerWebSocketCloseStatusDescription = statusDescription;
+                _innerWebSocket!.Invoke("close", (int)closeStatus, statusDescription);
+                await _tcsClose.Task.ConfigureAwait(continueOnCapturedContext: true);
```
`return _tcsClose.Task` instead

stephentoub replied:
> return _tcsClose.Task instead

Not if the intention is for the cancellation registration to persist until after this completes.

But I have the same concern here that I've voiced elsewhere: this pattern if calling TrySetCanceled in response to cancellation is only causing the caller to wake up, it's not actually canceling the underlying operation, and so the caller will think the operation has completed when it may not have.  In general that can lead to bad race conditions.

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    // Attach the onMessage callaback
+                    _innerWebSocket.SetObjectProperty("onmessage", _onMessage);
+
+                    await tcsConnect.Task.ConfigureAwait(continueOnCapturedContext: true);
```
`return tcsConnect.Task;`

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ClientWebSocketOptions.cs:
```diff
@@ -0,0 +1,117 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Do we need to specialize this file for the browser?  Or instead could ClientWebSocket.ConnectAsync just throw PNSE if it finds options set that aren't supported?

marek-safar replied:
The list of dependencies from the desktop version is much bigger (size savings) but it'd be nice to extract the common code to the shared file

stephentoub replied:
> The list of dependencies from the desktop version is much bigger (size savings)

Only if the consumer uses them, though, right?  Otherwise we should be able to ensure the relevant stuff is linked away?

marek-safar replied:
Yes if we don't call them internally, for example, to check if there were set.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+        {
+            if (State == WebSocketState.Open)
+            {
+                var closeResult = CloseAsyncCore(WebSocketCloseStatus.NormalClosure, "Connection was aborted", CancellationToken.None);
```
Presumably strings like "Connection was aborted" should be in the .resx?

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            if (messageType != WebSocketMessageType.Binary &&
+                messageType != WebSocketMessageType.Text)
+            {
+                throw new ArgumentException($"Invalid message type: '{messageType}' specified in method 'SendAsync'.  Valid types are 'Binary' and 'Text'",
```
Exception messages should be in the resx

kjpou1 replied:
Addressed

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully send or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            var writtenBuffer = _writeBuffer;
```
var => whatever type this is

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            {
+                _writeBuffer.Write(buffer.Array, buffer.Offset, buffer.Count);
+                if (!_writeBuffer.TryGetBuffer(out buffer))
+                    throw new WebSocketException(WebSocketError.NativeError);
```
You control how the MemoryStream is created, so this will never fail.  You can assert the result of TryGetBuffer instead.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
```
```suggestion
            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetCanceled(), tcsSend))
```

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    throw new WebSocketException(WebSocketError.NativeError);
+            }
+
+            var tcsSend = new TaskCompletionSource<bool>();
```
Same non-generic TaskCompletionSource comment.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    {
+                        using (var uint8Buffer = Uint8Array.From(buffer))
+                        {
+                            _innerWebSocket!.Invoke("send", uint8Buffer);
```
Is this synchronously sending the message?  Is it guaranteed to be fast?

kjpou1 replied:
Yes this is synchronously sending the message but there is not an async method.  The send should return immediately as it supposedly buffers data but that will depend on the implementation of the browser.

Guaranteed to be as fast as the browser implementation can accept, buffer and return.  But we can not really guarantee that the implantation is fast.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
```
I don't understand the purpose of this. This is going to cause the tcs to be canceled, but it's not actually preventing any of the sending work from happening, and it's only be awaited after all the work has been done.

kjpou1 replied:
Addressed in recent commits

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                {
+                    writtenBuffer?.Dispose();
+                }
+                await tcsSend.Task.ConfigureAwait(continueOnCapturedContext: true);
```
In what situation will this complete asynchronously?  It seems like the whole operation here is actually synchronous.

kjpou1 replied:
Addressed in recent commits

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully receive or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            using (cancellationToken.Register(() => tcsReceive.TrySetCanceled()))
```
I have the same questions here as I wrote for SendAsync.

kjpou1 replied:
addressed

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                {
+                    var endOfMessage = _bufferedPayload.BufferPayload(buffer, out WebSocketReceiveResult receiveResult);
+
+                    tcsReceive.SetResult(receiveResult);
```
This is going to throw an exception if cancellation occurs and completes the task first.

kjpou1 replied:
Stephen

The TCS is needed because we do not have threading enabled at this time.  The only way to have await work is to use TaskCompletionSource to control when the task should return with the result.

stephentoub replied:
That explanation doesn't make sense to me.  You have the receiveResult here.  Why can't you just `return receiveResult;`?

kjpou1 replied:
yes I see what you mean now.

kjpou1 replied:
Addressed in recent commit

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                        using (var uint8Buffer = Uint8Array.From(buffer))
+                        {
+                            _innerWebSocket!.Invoke("send", uint8Buffer);
+                            tcsSend.SetResult(true);
```
This is going to throw an exception if cancellation occurs and completes the task first.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    }
+                }
+            }
+        }
```
Can you use a `Channel.CreateUnbuffered<T>()` instead of this `ActionQueue<T>`?  Then you'd just use its WriteAsync/TryWrite instead of BufferPayload and its ReadAsync instead of DequeuePayloadAsync.

davidfowl replied:
I was just going to suggest a channel 😄

kjpou1 replied:
Yes, was going to look at all this Payload/ActionQueue implementation.  This was leftover from the mono implementation when all of this was implemented outside of the repo tree.  Now that it is incorporated and we can actually run tests against the implementation am quite sure this can all be refactored out.

kjpou1 replied:
Addressed in recent commits

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ReceivePayload.cs:
```diff
@@ -0,0 +1,53 @@
+            var bytesTransferred = Math.Min(_dataMessageReceived.Length - _unconsumedDataOffset, arraySegment.Count);
+            var endOfMessage = (_dataMessageReceived.Length - _unconsumedDataOffset) <= arraySegment.Count;
+
+            if (arraySegment.Array != null)
```
It is because of `error CS8604: Possible null reference argument for parameter`

stephentoub replied:
Yes, but this is only used internally.  Most of the callers are actually constructing the ArraySegment and are doing so in a way where it'll never be null.  For the WebSocket.ReceiveAsync public entrypoint method, that method needs to validate the ArraySegment with https://github.com/dotnet/runtime/blob/a7f2f43ec3cad3c5df596cedea6ab5cb221e73e5/src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs#L120, at which point this also won't be null.  So rather than doing a runtime check, we should just either Debug.Assert that it's not null, or use `!` on the Array, both of which will silence the warning.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

# Conflicts:
#	src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj
#	src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs
#	src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

kjpou1 committed:
Add ConnectAsync implementation to create a new instance of the BrowserWebSocket implementation.

kjpou1 committed:
Address review comments

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                    }
+                    _innerWebSocket.SetObjectProperty("binaryType", "arraybuffer");
+
+                    subProtocols?.Dispose();
```
This can now also go into the `requestedSubProtocols?.Count > 0` case (maybe as a `using`) because it is only used there.

kjpou1 replied:
yes agree.  Thanks

kjpou1 committed:
Address subprotocols comments

kjpou1 committed:
Remove internal custom class ActionQueue class in favor of using a Channel.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Fix object leak.

- Lambdas are not being released automatically.

kjpou1 committed:
Add doc for the constant that is being used.

- Address review comment

kjpou1 committed:
Clean up SendAsync method.

- Validate the message type.
- Validate the array segment that is passed as the message.

kjpou1 committed:
Add validation to ReceiveAsync for invalid buffer.

- remove unnecessary null checks from message buffering.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Address review comments

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ReceivePayload.cs:
```diff
@@ -0,0 +1,53 @@
+            if (arraySegment.Array != null)
+                Buffer.BlockCopy(_dataMessageReceived, _unconsumedDataOffset, arraySegment.Array, arraySegment.Offset, bytesTransferred);
+
+            if (!endOfMessage)
```
addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ReceivePayload.cs:
```diff
@@ -0,0 +1,53 @@
+        public bool BufferPayload(ArraySegment<byte> arraySegment, out WebSocketReceiveResult receiveResult)
+        {
+            var bytesTransferred = Math.Min(_dataMessageReceived.Length - _unconsumedDataOffset, arraySegment.Count);
+            var endOfMessage = (_dataMessageReceived.Length - _unconsumedDataOffset) <= arraySegment.Count;
```
addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/ReceivePayload.cs:
```diff
@@ -0,0 +1,53 @@
+
+        public bool BufferPayload(ArraySegment<byte> arraySegment, out WebSocketReceiveResult receiveResult)
+        {
+            var bytesTransferred = Math.Min(_dataMessageReceived.Length - _unconsumedDataOffset, arraySegment.Count);
```
addressed

kjpou1 committed:
Address review comments about TCS

kjpou1 committed:
Address camel case by using enum

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+        private const int _created = 0;
+        private const int _connecting = 1;
+        private const int _connected = 2;
+        private const int _disposed = 3;
```
addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+                {
+                    _created => WebSocketState.None,
+                    _connecting => WebSocketState.Connecting,
+                    _disposed => WebSocketState.Closed, // We only get here if disposed before connecting
```
addressed

kjpou1 committed:
Address abort request

kjpou1 committed:
Handle nullable

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            var tcsConnect = new TaskCompletionSource<bool>();
+
+            // For Abort/Dispose.  Calling Abort on the request at any point will close the connection.
+            _cts.Token.Register(AbortRequest);
```
addressed

kjpou1 committed:
Address review comment for removing cancellationtoken registration to the connect source.

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => tcsConnect.TrySetCanceled()))
```
addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => tcsConnect.TrySetCanceled()))
```
addressed by removing the dependency on the using

kjpou1 committed:
Use non-generic TaskCompletionSource

kjpou1 committed:
Cleanup var usage

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,529 @@
+
+        internal async Task ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List<string>? requestedSubProtocols)
+        {
+            var tcsConnect = new TaskCompletionSource<bool>();
```
thanks

kjpou1 committed:
Add string to .resx

kjpou1 committed:
Inline dispose

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Fix WebSocket opening exception

kjpou1 commented:
@marek-safar @stephentoub Hey guys could you guys give a quick once over while the other review comments are being addressed.  Working through some testing regressions.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Remove the asynchronous completion from SendAsync.

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,481 @@
+                    throw new WebSocketException(WebSocketError.NativeError);
+            }
+
+            TaskCompletionSource tcsSend = new TaskCompletionSource();
```
Unused.

kjpou1 replied:
Thanks

kjpou1 committed:
Fix object leak, exception on close and address review comments

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,478 @@
+                {
+                    using (evt)
+                    {
+                        if (!cancellationToken.IsCancellationRequested)
```
Should the `tcsConnect` be cancelled by this `cancellationToken` when `IsCancellationRequested` is `true`? Otherwise the `tcsConnect` never completes.

kjpou1 replied:
Addressed

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,478 @@
+                            if (Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != (int)InternalState.Connecting)
+                            {
+                                // Aborted/Disposed during connect.
+                                throw new ObjectDisposedException(GetType().FullName);
```
Should this exception be set on `tcsConnect` instead? Otherwise the `tcsConnect` never completes.

kjpou1 replied:
Addressed

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,478 @@
+                                // Aborted/Disposed during connect.
+                                throw new ObjectDisposedException(GetType().FullName);
+                            }
+                            tcsConnect.SetResult();
```
`onopen` will not be called if the JS `WebSocket` fails the handshake, instead only the `onerror` and `onclose` event will be called. Which means that the `tcsConnect` will never complete in this case.
Should `onerror` or `onclose` also try to set an exception on `tcsConnect`?

Example:
```js
var ws = new WebSocket("wss://github.com");
ws.onopen = e => console.log("open", e);
ws.onerror = e => console.log("error", e);
ws.onclose = e => console.log("close", e);
```
```
error Event {isTrusted: true, type: "error", target: WebSocket, currentTarget: WebSocket, eventPhase: 2, …}
close CloseEvent {isTrusted: true, wasClean: false, code: 1006, reason: "", type: "close", …}
```

kjpou1 replied:
Yes, that is the case.  Have not noticed that before.

kjpou1 replied:
Ah, I see now.  The code that set the tcsConnect result was removed from close callback.  I need to put it back.  Thanks for the catch.

kjpou1 replied:
This is now fixed.

kjpou1 committed:
Handle race condition

kjpou1 committed:
Address 	TaskCompletionSource on connect review comments

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,482 @@
+                                // Aborted/Disposed during connect.
+                                tcsConnect.SetException(new ObjectDisposedException(GetType().FullName));
+                            }
+                            tcsConnect.SetResult();
```
```suggestion
                            else
                            {
                                tcsConnect.SetResult();
                            }
```

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,482 @@
+                        }
+                        else
+                        {
+                            tcsConnect.SetCanceled();
```
```suggestion
                            tcsConnect.SetCanceled(cancellationToken);
```

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

kjpou1 committed:
Fix Connect not completing task on error.

kjpou1 committed:
Cleanup reference

kjpou1 committed:
Remove redundant error check of memorystream buffer.

kjpou1 committed:
cleanup

MaximLipnin commented on src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj:
```diff
@@ -35,5 +38,9 @@
     <Reference Include="System.Security.Cryptography.Primitives" />
+    <Reference Include="System.Threading.Channels" Condition=" '$(TargetsBrowser)' == 'true'" />
   </ItemGroup>
+  <ItemGroup Condition=" '$(TargetsBrowser)' == 'true'" >
```
```suggestion
  <ItemGroup Condition="'$(TargetsBrowser)' == 'true'" >
```

MaximLipnin commented on src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj:
```diff
@@ -35,5 +38,9 @@
+    <Reference Include="System.Threading.Channels" Condition=" '$(TargetsBrowser)' == 'true'" />
   </ItemGroup>
+  <ItemGroup Condition=" '$(TargetsBrowser)' == 'true'" >
+    <ProjectReference Include="..\..\System.Runtime.InteropServices.JavaScript/src/System.Runtime.InteropServices.JavaScript.csproj" />
```
```suggestion
    <ProjectReference Include="..\..\System.Runtime.InteropServices.JavaScript\src\System.Runtime.InteropServices.JavaScript.csproj" />
```

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj

Co-authored-by: Maxim Lipnin <mlipnin@gmail.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj

Co-authored-by: Maxim Lipnin <mlipnin@gmail.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj

Co-authored-by: Maxim Lipnin <mlipnin@gmail.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj

Co-authored-by: Maxim Lipnin <mlipnin@gmail.com>

kjpou1 committed:
Fix typo

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,490 @@
+                                                    using (ArrayBuffer binResult = (ArrayBuffer)target.GetObjectProperty("result"))
+                                                    {
+                                                        _receiveMessageQueue.Writer.TryWrite(new ReceivePayload(binResult, WebSocketMessageType.Binary));
+                                                        if (loadend != null)
+                                                            JavaScript.Runtime.FreeObject(loadend);
```
I don't understand this part, which object are you trying to release?

kjpou1 replied:
Cleaned up deprecated pre ref counting code.

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,490 @@
+                _onMessage = new Action<JSObject>((messageEvent) =>
+                {
+                    // get the events "data"
+                    using (messageEvent)
```
Could you extract this lambda into normal method?

kjpou1 replied:
Addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,490 @@
+
+            if (!endOfMessage)
+            {
+                _writeBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
```
What about moving this code before the `if`

kjpou1 replied:
Actually thought I had.

Addressed

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,490 @@
+
+            WebSocketValidate.ValidateArraySegment(buffer, nameof(buffer));
+
+            _writeBuffer ??= new MemoryStream();
```
Is this intentional? Caling SendAsync repeatably will grow MemoryStream with no limit

kjpou1 replied:
The write buffer is nulled out below so it will be recreated.

marek-safar replied:
then why do you need `??=`

kjpou1 replied:
because the message needs to be buffered until end of message is reached.  Once end of message is reached it gets nulled out and needs to be recreated.

kjpou1 committed:
clean up buffer code

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets

kjpou1 committed:
Modify callbacks to use lambda function.

- These are now released properly after reference counting went in.

kjpou1 committed:
Fix object leak of delegate and clean up deprecated code.

kjpou1 committed:
extract lambda into method

marek-safar reviewed (approved):
LGTM from me. @stephentoub could you do another pass

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,483 @@
+            }
+            catch (Exception excb)
+            {
+                throw new WebSocketException(WebSocketError.NativeError, excb);
```
```suggestion
                return new Task.FromException(WebSocketException(WebSocketError.NativeError, excb));
```

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,483 @@
+            _writeBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
+
+            if (!endOfMessage)
+                return Task.CompletedTask;
```
The browser websocket API isn't capable of sending messages split over multiple fragments?  The `endOfMessage` argument elsewhere translates into the FIN bit in the websocket frame.

kjpou1 replied:
We do not have access to the websocket frame through the browser api.  Also, from my understanding there is not the concept of EndOfMessage.  Not that I can find anyway.

lewing replied:
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send expects the full message so we need to buffer somewhere.  An alternative to the MemoryStream buffer would be constructing a chain of Blobs on the JS side and calling send on that when endOfMessage is true.

stephentoub replied:
It's fine for the buffering to be in managed.  I was questioning the buffering all-up, but if the websocket API doesn't support fragments, then ok.

stephentoub commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,483 @@
+
+        private async Task CloseAsyncCore(WebSocketCloseStatus closeStatus, string? statusDescription, CancellationToken cancellationToken)
+        {
+            ThrowOnInvalidState(State,
+                WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
```
Nit:
```suggestion
            ThrowOnInvalidState(State, WebSocketState.Open, WebSocketState.CloseReceived, WebSocketState.CloseSent);
```

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

kjpou1 committed:
Update src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

campersau commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,479 @@
+            }
+            catch (Exception excb)
+            {
+                return new Task.FromException(WebSocketException(WebSocketError.NativeError, excb));
```
I think that's what @stephentoub meant:

```suggestion
                return Task.FromException(new WebSocketException(WebSocketError.NativeError, excb));
```

stephentoub replied:
Yup, typo.

kjpou1 replied:
Yes, I already updated it.  Thanks

kjpou1 committed:
Review comments addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,483 @@
+                            if (Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != (int)InternalState.Connecting)
+                            {
+                                // Aborted/Disposed during connect.
+                                tcsConnect.SetException(new ObjectDisposedException(GetType().FullName));
```
addressed

kjpou1 commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs:
```diff
@@ -0,0 +1,483 @@
+                        NativeCleanup();
+                        if ((InternalState)_state == InternalState.Connecting)
+                        {
+                            tcsConnect.SetException(new WebSocketException(WebSocketError.NativeError));
```
addressed
